### ISSUE TITLE
Well-organized benchmarking suite

### ISSUE THREAD

**Issue by uellue** [LiberTEM/LiberTEM#198]
Right now, we do have automated tests, but we don't have automated benchmarks. 

We should have a benchmarking suite that can be executed to gauge the impact of a change. The unit tests are not really suitable because they often use LiberTEM in a non-standard way: inline executor instead of dask, very small in-memory data sets, test of "weird" partitioning schemes, using sparse matrices where one really shouldn't #182 .

Instead, we should have a reliable test program with the following properties:

* Test on large real-world data sets with typical analyses
* Executed under reproducible conditions: Dedicated machine without background tasks
* Executed on a small cluster as well as a typical single-node workstation.
* Collect profiling information
* Regular automated collection and reporting of benchmark results to catch changes with a negative impact on performance.

**Comment by sk1p** [comment]
See also: https://github.com/airspeed-velocity/asv

**Comment by uellue** [comment]
The benchmarking suite should separate processing and overheads. Overheads can be a major factor #232

**Comment by uellue** [comment]
We should routinely compare math libraries:

* Torch vs openblas (vs mkl?)
* pyfftw vs numpy.fft vs scipy.fft (vs mkl?)

As a side note, it seems that openblas might have caught up to torch recently -- further tests desirable.

**Comment by sk1p** [comment]
Btw. for FFT comparison, see also https://www.kcod.de/gearshifft/

**Comment by sk1p** [comment]
Idea: use an Azure on-premise Agent to run the benchmarks as part of the CI pipeline, on a dedicated server. Start with tests that can run in a few seconds, maybe later introduce daily benchmark run for longer running tests (for example ones that create some TB of data on demand, also to test our writing facilities - things you don't want to run on every commit)

**Comment by uellue** [comment]
Asvlooks like the best bet so far and should cover all points when used appropriately.

A few more detailed thoughts on use cases and intermediate results after playing with asv:

* Routinely check during development if a change introduced regressions to catch things like #814 before merging
  - That means asv has to run code from current fork, current branch.
  - Using the "same repo" layout collides with the existing structures of LiberTEM.
  - Using the "separate repo" layout in a separate repo LiberTEM/benchmarks required setting a Git URL or path to the actual repo to benchmark, meaning one would have to merge into main repo master first or adapt the path to the way how the repos are checked out, which is fragile.
  - Using the "separate repo" layout in a subfolder of the main repo, and setting the repo URL/path to a relative path pointing to the repository root ('..') worked.
* Many benchmarks and required fixtures are equivalent to existing unit tests
  - We should make sure to re-use as much code as possible for both to avoid duplication
  - In particular the files for dataset tests are interesting!
  - The fixture mechanism and code loading of pytest is rather specific and I couldn't find documentation on how to use fixtures and run individual test functions without running the entire pytest suite. Calling a fixture directly fails complaining loudly: https://docs.pytest.org/en/latest/deprecations.html#calling-fixtures-directly
  - In an ideal world, the functionality of asv to record and track timings as a function of commit, machine, config etc would be folded into pytest...
  - As a way out, perhaps one can define common utility functions in such a way that they can work within the asv benchmarks and wrap them into fixtures for pytest? Import or wrap some common test functions from that code base rather than defining them in the tests folder?
    + What would be a good place for common infrastructure? Within module of libertem package (easy to import), or repo subfolder (keep infra out of release, but harder to import)?
    +  Some functions may actually fit well in a separate repo (LiberTEM-infra?) since LiberTEM sub-packages like blobfinder duplicate them as well.
* We'll have to set up gitignore rules to mask out a lot of the intermediate data of asv (envs, HTML, ...)
* We'll have to figure out a process when to commit which parts of the asv directory structure to save benchmarks, and how to discard unwanted ones, making sure they are not committed.
* We'll have to think how to integrate the benchmarks in the gh-pages structure. Fortunately it seems that the html, css, js output of asv is relocatable -- to be verified!

**Comment by uellue** [comment]
Specific item: Test if the resource check and other parts of the the DaskJobExecutor scale to at least 1000 workers, better 10000.

**Comment by uellue** [comment]
Issue: Separate first run from repeat runs (Numba compile time etc)!

**Comment by uellue** [comment]
This one actually extends pytest and checks many of the items as well: https://pytest-benchmark.readthedocs.io/en/latest/index.html

**Comment by sk1p** [comment]
> This one actually extends pytest and checks many of the requirements as well: https://pytest-benchmark.readthedocs.io/en/latest/index.html

Nice! As I understand, this is missing the reporting functionality of airspeed velocity - if you could write up how that should work I could write a prototype for that.

**Comment by sk1p** [comment]
Or is plotting using `--benchmark-histogram` enough for our purposes?

**Comment by uellue** [comment]
Yes, it only has basic reporting. Since the data is available as JSON, it would not be too difficult to write a custom reporting function that discovers the files, extracts the relevant data and plots it, for example with matplotlib. We could call that as part of the build process like we do with `build-authors-contributors`.

**Comment by uellue** [comment]
...cont: Advantage would be that we have full control what is plotted how and how tests are grouped, including extracting the first run vs. repeats, overlay multiple plots for comparison, custom units such as pix/s, etc. Disadvantage is of course more work to make it look nice. IMO a trade-off between cost of developing and maintaining benchmarks (lower with pytest) vs. cost of reporting.

**Comment by sk1p** [comment]
As #565 is merged, we could now create a pipeline to run this regularly. As load on our on-prem agent varies, we would have to see what variances are introduced, and how well the iterations/rounds etc stabilizes the numbers.
